File: codegen/templates/csharp/dynamic/BaseAI.cs
using System;
using System.Runtime.InteropServices;

/// <summary>
/// This class implements most the code an AI would need to interface with the lower-level game code.
/// AIs should extend this class to get a lot of builer-plate code out of the way. The provided AI class does just that.
/// </summary>
public abstract class BaseAI
{
% for model in models:
%   if model.type == 'Model':
  public static ${model.name}[] ${lowercase(model.plural)};
%   endif
% endfor

  IntPtr connection;
  public static int iteration;
  bool initialized;

  public BaseAI(IntPtr c)
  {
    connection = c;
  }

  /// <summary>
  /// Make this your username, which should be provided.
  /// </summary>
  /// <returns>Returns the username of the player.</returns>
  public abstract String username();

  /// <summary>
  /// Make this your password, which should be provided.
  /// </summary>
  /// <returns>Returns the password of the player.</returns>
  public abstract String password();

  /// <summary>
  /// This is run once on turn one before run().
  /// </summary>
  public abstract void init();

  /// <summary>
  /// This is run every turn.
  /// </summary>
  /// <returns>
  /// Return true to end turn, false to resynchronize with the 
  /// server and run again.
  /// </returns>
  public abstract bool run();

  /// <summary>
  /// This is run once after your last turn.
  /// </summary>
  public abstract void end();

  /// <summary>
  /// Synchronizes with the server, then calls run().
  /// </summary>
  /// <returns>
  /// Return true to end turn, false to resynchronize with the 
  /// server and run again.
  /// </returns>
  public bool startTurn()
  {
    int count = 0;
    iteration++;

% for model in models:
%   if model.type == 'Model':
    count = Client.get${model.name}Count(connection);
    ${lowercase(model.plural)} = new ${model.name}[count];
    for(int i = 0; i < count; i++)
      ${lowercase(model.plural)}[i] = new ${model.name}(Client.get${model.name}(connection, i));

%   endif
% endfor
    if(!initialized)
    {
      initialized = true;
      init();
    }

    return run();
  }
% for datum in globals:

  /// <summary>
  /// ${datum.doc}
  /// </summary>
  /// <returns>Returns ${lowercase(datum.doc)}</returns>
  public ${types[datum.type]} ${datum.name}()
  {
    ${types[datum.type]} value = Client.get${capitalize(datum.name)}(connection);
%   if datum.type is str:
    return Marshal.PtrToStringAuto(value);
%   else:
    return value;
%   endif
  }
% endfor
}

File: testClients/StephenJ/BaseAI.py
# -*- python -*-

from library import library

class BaseAI:
  """@brief A basic AI interface.

  This class implements most the code an AI would need to interface with the lower-level game code.
  AIs should extend this class to get a lot of builer-plate code out of the way
  The provided AI class does just that.
  """
  #\cond
  initialized = False
  iteration = 0
  runGenerator = None
  connection = None
  #\endcond
  players = []
  mappables = []
  droids = []
  tiles = []
  modelVariants = []
  #\cond
  def startTurn(self):
    from GameObject import Player
    from GameObject import Mappable
    from GameObject import Droid
    from GameObject import Tile
    from GameObject import ModelVariant

    BaseAI.players = [Player(library.getPlayer(self.connection, i)) for i in xrange(library.getPlayerCount(self.connection))]
    BaseAI.mappables = [Mappable(library.getMappable(self.connection, i)) for i in xrange(library.getMappableCount(self.connection))]
    BaseAI.droids = [Droid(library.getDroid(self.connection, i)) for i in xrange(library.getDroidCount(self.connection))]
    BaseAI.tiles = [Tile(library.getTile(self.connection, i)) for i in xrange(library.getTileCount(self.connection))]
    BaseAI.modelVariants = [ModelVariant(library.getModelVariant(self.connection, i)) for i in xrange(library.getModelVariantCount(self.connection))]

    if not self.initialized:
      self.initialized = True
      self.init()
    BaseAI.iteration += 1;
    if self.runGenerator:
      try:
        return self.runGenerator.next()
      except StopIteration:
        self.runGenerator = None
    r = self.run()
    if hasattr(r, '__iter__'):
      self.runGenerator = r
      return r.next()
    return r
  #\endcond
  #\cond
  def getMapWidth(self):
    return library.getMapWidth(self.connection)
  #\endcond
  mapWidth = property(getMapWidth)
  #\cond
  def getMapHeight(self):
    return library.getMapHeight(self.connection)
  #\endcond
  mapHeight = property(getMapHeight)
  #\cond
  def getTurnNumber(self):
    return library.getTurnNumber(self.connection)
  #\endcond
  turnNumber = property(getTurnNumber)
  #\cond
  def getMaxDroids(self):
    return library.getMaxDroids(self.connection)
  #\endcond
  maxDroids = property(getMaxDroids)
  #\cond
  def getPlayerID(self):
    return library.getPlayerID(self.connection)
  #\endcond
  playerID = property(getPlayerID)
  #\cond
  def getGameNumber(self):
    return library.getGameNumber(self.connection)
  #\endcond
  gameNumber = property(getGameNumber)
  #\cond
  def getScrapRate(self):
    return library.getScrapRate(self.connection)
  #\endcond
  scrapRate = property(getScrapRate)
  #\cond
  def getMaxScrap(self):
    return library.getMaxScrap(self.connection)
  #\endcond
  maxScrap = property(getMaxScrap)
  #\cond
  def getDropTime(self):
    return library.getDropTime(self.connection)
  #\endcond
  dropTime = property(getDropTime)
  def __init__(self, connection):
    self.connection = connection
